node-js-basics-101/node_modules/asynquence-contrib/plugin.goCSP.js

```diff
@@ -0,0 +1,573 @@
+			ret.val(cb);
+		}
+		else {
+			return ret;
+		}
+	}
+
+	function takem(channel) {
+		var msg;
+
+		if (channel.closed) {
+			return ASQ.csp.CLOSED;
+		}
+
+		// remove already-resolved entries
+		channel.put_queue = filterResolved(channel.put_queue);
+		channel.take_queue = filterResolved(channel.take_queue);
+
+		// immediate take?
+		if (channel.messages.length > 0) {
+			msg = channel.messages.shift();
+			unblock(channel.put_queue.shift());
+			if (msg instanceof Error) {
+				throw msg;
+			}
+			return msg;
+		}
+		// queued take
+		else {
+			channel.take_queue.push(
+				// make a notifiable iterable for 'take' blocking
+				ASQ.iterable()
+				.then(function $$then(){
+					if (!channel.closed) {
+						var v = channel.messages.shift();
+						if (v instanceof Error) {
+							throw v;
+						}
+						return v;
+					}
+					else {
+						return ASQ.csp.CLOSED;
+					}
+				})
+			);
+
+			// wrap a sequence/promise around the iterable
+			msg = ASQ(
+				channel.take_queue[channel.take_queue.length - 1]
+			);
+
+			// put waiting on this take?
+			if (channel.put_queue.length > 0) {
+				unblock(channel.put_queue.shift());
+				unblock(channel.take_queue.shift());
+			}
+
+			return msg;
+		}
+	}
+
+	function takemAsync(channel,cb) {
+		var ret = ASQ(takem(channel));
+
+		if (cb && typeof cb == "function") {
+			ret.pThen(cb,cb);
+		}
+		else {
+			return ret.val(function $$val(v){
+				if (v instanceof Error) {
+					throw v;
+				}
+				return v;
+			});
+		}
+	}
+
+	function alts(actions) {
+		var closed, open, handlers, i, isq, ret, resolved = false;
+
+		// used `alts(..)` incorrectly?
+		if (!Array.isArray(actions) || actions.length == 0) {
+			throw Error("Invalid usage");
+		}
+
+		closed = [];
+		open = [];
+		handlers = [];
+
+		// separate actions by open/closed channel status
+		actions.forEach(function $$each(action){
+			var channel = Array.isArray(action) ? action[0] : action;
+
+			// remove already-resolved entries
+			channel.put_queue = filterResolved(channel.put_queue);
+			channel.take_queue = filterResolved(channel.take_queue);
+
+			if (channel.closed) {
+				closed.push(channel);
+			}
+			else {
+				open.push(action);
+			}
+		});
+
+		// if no channels are still open, we're done
+		if (open.length == 0) {
+			return { value: ASQ.csp.CLOSED, channel: closed };
+		}
+
+		// can any channel action be executed immediately?
+		for (i=0; i<open.length; i++) {
+			// put action
+			if (Array.isArray(open[i])) {
+				// immediate put?
+				if (open[i][0].messages.length < open[i][0].buffer_size) {
+					return { value: put(open[i][0],open[i][1]), channel: open[i][0] };
+				}
+			}
+			// immediate take?
+			else if (open[i].messages.length > 0) {
+				return { value: take(open[i]), channel: open[i] };
+			}
+		}
+
+		isq = ASQ.iterable();
+		var ret = ASQ(isq);
+
+		// setup channel action handlers
+		for (i=0; i<open.length; i++) {
+			(function iteration(action,channel,value){
```

Comment:
__[jshint]__
* Functions declared within loops referencing an outer scoped variable may lead to confusing semantics.

<!-- Comment by [SideCI](https://sideci.com) -->